src/GafferImage/ImageStats.cpp

```diff
@@ -198,98 +222,219 @@ const DeepState *ImageStats::deepState() const
-	Sampler s( inPlug(), channelName, area );
+	if( output == tileStatsPlug() )
+	{
+		Imath::V2i tileOrigin = context->get<Imath::V2i>( ImagePlug::tileOriginContextName );
+		const Imath::Box2i tileBound = BufferAlgo::intersection(
+			Imath::Box2i( boundsIntersection.min - tileOrigin, boundsIntersection.max - tileOrigin ),
+			Imath::Box2i( Imath::V2i( 0 ), Imath::V2i( ImagePlug::tileSize() ) )
+		);
 
-	float min = Imath::limits<float>::max();
-	float max = Imath::limits<float>::min();
-	double sum = 0.;
+		IECore::ConstFloatVectorDataPtr channelData = flattenedInPlug()->channelDataPlug()->getValue();
 
-	for( int y = area.min.y; y < area.max.y; ++y )
-	{
-		for( int x = area.min.x; x < area.max.x; ++x )
+		float min = Imath::limits<float>::max();
+		float max = Imath::limits<float>::min();
+		double sum = 0.;
+
+		const std::vector<float> &channel = channelData->readable();
+		for( int y = tileBound.min.y; y < tileBound.max.y; ++y )
 		{
-			float v = s.sample( x, y );
-			min = std::min( v, min );
-			max = std::max( v, max );
-			sum += v;
+			for( int x = tileBound.min.x; x < tileBound.max.x; ++x )
+			{
+				float v = channel[ x + y * ImagePlug::tileSize() ];
+				min = std::min( v, min );
+				max = std::max( v, max );
+				sum += v;
+			}
 		}
-	}
 
-	if( output->parent<Plug>() == minPlug() )
-	{
-		static_cast<FloatPlug *>( output )->setValue( min );
+		// Note that we're converting the sum to float for storage.  I don't see this as being
+		// a problem - we're going to be dividing this value down at the end.  I suppose it
+		// would hypothetically be a problem in an image with large regions that were extremely
+		// negative, cancelling out with positive regions, to produce a small average, but this
+		// seems like an unlikely case.
+		static_cast<AtomicBox3fPlug *>( output )->setValue( Imath::Box3f( Imath::V3f( min, max, sum ), Imath::V3f( 0 ) ) );
 	}
-	else if( output->parent<Plug>() == maxPlug() )
+	else if( output == allStatsPlug() )
 	{
-		static_cast<FloatPlug *>( output )->setValue( max );
-	}
-	else if( output->parent<Plug>() == averagePlug() )
-	{
-		static_cast<FloatPlug *>( output )->setValue(
-			sum / double( (area.size().x) * (area.size().y) )
+		if( BufferAlgo::empty( boundsIntersection ) )
+		{
+			static_cast<AtomicBox3fPlug *>( output )->setValue( Imath::Box3f( Imath::V3f( 0 ), Imath::V3f( 0 ) ) );
+			return;
+		}
+		float min = Imath::limits<float>::max();
+		float max = Imath::limits<float>::min();
+		double sum = 0.;
+		ImageAlgo::parallelGatherTiles(
+			flattenedInPlug(),
+			// Tile
+			[this] ( const ImagePlug *imageP, const Imath::V2i &tileOrigin )
+			{
+				return tileStatsPlug()->getValue();
+			},
+			// Gather
+			[ &min, &max, &sum ] ( const ImagePlug *imageP, const Imath::V2i &tileOrigin, const Imath::Box3f &v )
+			{
+				min = std::min( v.min[0], min );
+				max = std::max( v.min[1], max );
+				sum += v.min[2];
+			},
+			boundsIntersection,
+			ImageAlgo::BottomToTop
```

Comment:
Why BottomToTop?

Reply:
From separate discussion : should have been `TopToBottom`, but we do need to avoid `Unordered`. For hash, we need a deterministic order to have a repeatable hash. For compute, floating point results would differ with different orders of addition.